In [2]:
# importing libraries
import pandas as pd
import numpy as np
import os

In [21]:
path = r'C:\Users\ali\OneDrive\Desktop\careerfoundry\Data Immersion\Achievement 4\15-08-2025 Instacart Basket Analysis'
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'ords_prods_merge_innerjoin.pkl'))

In [22]:
ords_prods_merge.shape

(32404289, 14)

In [4]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  is_first_order  product_id  add_to_cart_order  \
0                     NaN               1         196                  1   
1                     NaN               1       14084                  2   
2                     NaN               1       12427                  3   
3                     NaN               1       26088                  4   
4                     NaN               1       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices  
0              7     9.0  
1             16    12.5  
2             19     4.4  
3             19     4.7  
4             17     1.0

In [5]:
# creating subset of ords_prods_merge 
df = ords_prods_merge[:1000000]

In [6]:
df.shape

(1000000, 14)

## Defining a function

In [7]:
# define a function
def price_label(row):

  if row['prices'] <= 5:
    return 'Low-range product'
  elif (row['prices'] > 5) and (row['prices'] <= 15):
    return 'Mid-range product'
  elif row['prices'] > 15:
    return 'High range'
  else: return 'Not enough data'

In [9]:
# applying function on df dataframe
df['price_range'] = df.apply(price_label, axis=1)

C:\Users\muham\AppData\Local\Temp\ipykernel_29224\1060633564.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df['price_range'] = df.apply(price_label, axis=1)


In [10]:
df['price_range'].value_counts(dropna = False)

price_range
Mid-range product    673176
Low-range product    314399
High range            12425
Name: count, dtype: int64

In [11]:
df['prices'].max()

99999.0

## IF-statement with the loc() function

In [12]:
# IF-statement with the loc() function
df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'

C:\Users\muham\AppData\Local\Temp\ipykernel_29224\521822940.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.loc[df['prices'] > 15, 'price_range_loc'] = 'High-range product'


In [13]:
df.loc[(df['prices'] <= 15) & (df['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [14]:
df.loc[df['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [15]:
df['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     673176
Low-range product     314399
High-range product     12425
Name: count, dtype: int64

### Handling outliers(max value 99999)

In [5]:
ords_prods_merge.loc[ords_prods_merge['prices'] > 100]

order_id  user_id  order_number  orders_day_of_week  \
1576        912404       17            12                   2   
1638        603376       17            22                   6   
16534      3264360      135             2                   2   
16540       892534      135             3                   0   
53711       229704      342             8                   1   
...            ...      ...           ...                 ...   
32350064   3172853   205650            18                   1   
32377297   2504315   205818             3                   5   
32377307   1108388   205818             5                   4   
32410137   1916142   206049             1                   2   
32410161    379732   206049             4                   1   

          order_hour_of_day  days_since_prior_order  is_first_order  \
1576                     14                     5.0               0   
1638                     16                     4.0               0   
16534                    21                    13.0               0   
16540                     8                    12.0               0   
53711                    19                    30.0               0   
...                     ...                     ...             ...   
32350064                  9                     7.0               0   
32377297                 15                     3.0               0   
32377307                  5                     1.0               0   
32410137                 17                     NaN               1   
32410161                 14                     5.0               0   

          product_id  add_to_cart_order  reordered  \
1576           21553                  5          0   
1638           21553                  3          1   
16534          21553                  6          0   
16540          21553                  3          1   
53711          21553                  9          0   
...              ...                ...        ...   
32350064       21553                 17          1   
32377297       21553                 13          0   
32377307       21553                  5          1   
32410137       21553                  2          0   
32410161       21553                  4          1   

                              product_name  aisle_id  department_id   prices  
1576      Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  
1638      Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  
16534     Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  
16540     Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  
53711     Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  
...                                    ...       ...            ...      ...  
32350064  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  
32377297  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  
32377307  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  
32410137  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  
32410161  Lowfat 2% Milkfat Cottage Cheese       108             16  14900.0  

[5127 rows x 14 columns]

In [29]:
# replacing odd values/extra ordinary high values like 99999 with NaN for now
ords_prods_merge.loc[ords_prods_merge['prices'] >100, 'prices'] = np.nan

In [53]:
ords_prods_merge['prices'].max()

25.0

In [37]:
# using loc on full data frame
ords_prods_merge.loc[ords_prods_merge['prices'] > 15, 'price_range_loc'] = 'High-range product'

In [38]:
ords_prods_merge.loc[(ords_prods_merge['prices'] <= 15) & (ords_prods_merge['prices'] > 5), 'price_range_loc'] = 'Mid-range product' 

In [39]:
ords_prods_merge.loc[ords_prods_merge['prices'] <= 5, 'price_range_loc'] = 'Low-range product'

In [57]:
ords_prods_merge.loc[ords_prods_merge['prices'].isna(), 'price_range_loc'] = np.nan

In [58]:
ords_prods_merge['price_range_loc'].value_counts(dropna = False)

price_range_loc
Mid-range product     21860290
Low-range product     10126321
High-range product      412551
NaN                       5127
Name: count, dtype: int64

## If-Statements with For-Loops

In [59]:
# busiest day of week 
ords_prods_merge['orders_day_of_week'].value_counts(dropna = False)

orders_day_of_week
0    6204040
1    5660135
6    4496403
2    4213760
5    4205721
3    3840476
4    3783754
Name: count, dtype: int64

In [60]:
result = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0:
    result.append("Busiest day")
  elif value == 4:
    result.append("Least busy")
  else:
    result.append("Regularly busy")

In [ ]:
result

In [61]:
ords_prods_merge['busiest_day']= result

In [62]:
ords_prods_merge['busiest_day'].value_counts(dropna = False)

busiest_day
Regularly busy    22416495
Busiest day        6204040
Least busy         3783754
Name: count, dtype: int64

### 4.7 TASK

#### 2.

In [63]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  is_first_order  product_id  add_to_cart_order  \
0                     NaN               1         196                  1   
1                     NaN               1       14084                  2   
2                     NaN               1       12427                  3   
3                     NaN               1       26088                  4   
4                     NaN               1       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices    price_range_loc     busiest_day  
0              7     9.0  Mid-range product  Regularly busy  
1             16    12.5  Mid-range product  Regularly busy  
2             19     4.4  Low-range product  Regularly busy  
3             19     4.7  Low-range product  Regularly busy  
4             17     1.0  Low-range product  Regularly busy

In [64]:
#Defining a function to create a new column categorizing days of the week 
new_column = []

for value in ords_prods_merge["orders_day_of_week"]:
  if value == 0 or value == 1:
    new_column.append("Busiest days")
  elif value == 3 or value == 4:
    new_column.append("Slowest days")
  else:
    new_column.append("Regularly busy")

In [65]:
ords_prods_merge['busiest_days']= new_column

In [66]:
ords_prods_merge['busiest_days'].value_counts(dropna = False)

busiest_days
Regularly busy    12915884
Busiest days      11864175
Slowest days       7624230
Name: count, dtype: int64

In [67]:
ords_prods_merge = ords_prods_merge.drop('busiest_day', axis=1) 

In [28]:
ords_prods_merge.tail()

order_id  user_id  order_number  orders_day_of_week  \
32434484   2977660   206209            13                   1   
32434485   2977660   206209            13                   1   
32434486   2977660   206209            13                   1   
32434487   2977660   206209            13                   1   
32434488   2977660   206209            13                   1   

          order_hour_of_day  days_since_prior_order  is_first_order  \
32434484                 12                     7.0               0   
32434485                 12                     7.0               0   
32434486                 12                     7.0               0   
32434487                 12                     7.0               0   
32434488                 12                     7.0               0   

          product_id  add_to_cart_order  reordered  \
32434484       14197                  5          1   
32434485       38730                  6          0   
32434486       31477                  7          0   
32434487        6567                  8          0   
32434488       22920                  9          0   

                                   product_name  aisle_id  department_id  \
32434484                           Tomato Paste         9              9   
32434485        Brownie Crunch High Protein Bar         3             19   
32434486  High Protein Bar Chunky Peanut Butter         3             19   
32434487    Chocolate Peanut Butter Protein Bar         3             19   
32434488    Roasted & Salted Shelled Pistachios       117             19   

          prices    price_range_loc  busiest_days  
32434484     5.6  Mid-range product  Busiest days  
32434485     5.9  Mid-range product  Busiest days  
32434486     4.2  Low-range product  Busiest days  
32434487     4.9  Low-range product  Busiest days  
32434488     1.7  Low-range product  Busiest days

#### 03.

#### Both the columns busiest_day and busiest_days have the same total row count that is 32404289 which indicates that all the rows have been allocated a value based on day of the week.

#### 04.

In [80]:
# busiest hours of the day
hour_counts = ords_prods_merge['order_hour_of_day'].value_counts(dropna = False).sort_index()

In [84]:
hour_counts


order_hour_of_day
0      218766
1      115699
2       69374
3       51281
4       53241
5       87959
6      290492
7      891040
8     1718100
9     2454165
10    2761710
11    2736075
12    2618481
13    2660900
14    2689086
15    2662094
16    2535154
17    2087609
18    1636469
19    1258290
20     976145
21     795628
22     634216
23     402315
Name: count, dtype: int64

In [78]:
#using the 33rd and 66th percentiles to break into 3 groups
q33 = hour_counts.quantile(0.33)
q66 = hour_counts.quantile(0.66)

In [79]:
# Defining a function to assign category
def label_order_volume(count):
    if count <= q33:
        return 'Fewest Orders'
    elif count <= q66:
        return 'Average Orders'
    else:
        return 'Most Orders'

In [81]:
# Applying our function to hour_counts
hour_volume_labels = hour_counts.apply(label_order_volume)

In [82]:
# Mapping labels to original data frame
ords_prods_merge['busiest_period_of_day'] = ords_prods_merge['order_hour_of_day'].map(hour_volume_labels)

In [83]:
ords_prods_merge.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2539329        1             1                   2                  8   
2   2539329        1             1                   2                  8   
3   2539329        1             1                   2                  8   
4   2539329        1             1                   2                  8   

   days_since_prior_order  is_first_order  product_id  add_to_cart_order  \
0                     NaN               1         196                  1   
1                     NaN               1       14084                  2   
2                     NaN               1       12427                  3   
3                     NaN               1       26088                  4   
4                     NaN               1       26405                  5   

   reordered                             product_name  aisle_id  \
0          0                                     Soda        77   
1          0  Organic Unsweetened Vanilla Almond Milk        91   
2          0                      Original Beef Jerky        23   
3          0               Aged White Cheddar Popcorn        23   
4          0         XL Pick-A-Size Paper Towel Rolls        54   

   department_id  prices    price_range_loc    busiest_days  \
0              7     9.0  Mid-range product  Regularly busy   
1             16    12.5  Mid-range product  Regularly busy   
2             19     4.4  Low-range product  Regularly busy   
3             19     4.7  Low-range product  Regularly busy   
4             17     1.0  Low-range product  Regularly busy   

  busiest_period_of_day  
0        Average Orders  
1        Average Orders  
2        Average Orders  
3        Average Orders  
4        Average Orders

#### 05.

In [86]:
ords_prods_merge['busiest_period_of_day'].value_counts(dropna = False)

busiest_period_of_day
Most Orders       21117665
Average Orders     9997497
Fewest Orders      1289127
Name: count, dtype: int64

In [87]:
# Exporting data frame
ords_prods_merge.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ords_prods_merge_updated.pkl'))